Description: Do not retire the fallback DCID if it is still used
Author: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Origin: upstream, https://github.com/ngtcp2/ngtcp2/commit/1ffa646cb33ad8e7e64d21d97fa90205e6c6635d
Bug: https://github.com/ngtcp2/ngtcp2/pull/2286
Last-Update: 2026-10-05
---
diff --git a/lib/ngtcp2_conn.c b/lib/ngtcp2_conn.c
index 12cca44b..500afcd9 100644
--- a/lib/ngtcp2_conn.c
+++ b/lib/ngtcp2_conn.c
@@ -5077,7 +5077,10 @@ static int conn_stop_pv(ngtcp2_conn *conn, ngtcp2_tstamp ts) {
     return 0;
   }

-  if (pv->dcid.cid.datalen && pv->dcid.seq != conn->dcid.current.seq) {
+  if (pv->dcid.cid.datalen && pv->dcid.seq != conn->dcid.current.seq &&
+      (!(pv->flags & NGTCP2_PV_FLAG_FALLBACK_PRESENT) ||
+       !(pv->flags & NGTCP2_PV_FLAG_DONT_RETIRE_FALLBACK) ||
+       pv->dcid.seq != pv->fallback_dcid.seq)) {
     rv = conn_retire_active_dcid(conn, &pv->dcid, ts);
     if (rv != 0) {
       goto fin;
@@ -5085,6 +5088,7 @@ static int conn_stop_pv(ngtcp2_conn *conn, ngtcp2_tstamp ts) {
   }

   if ((pv->flags & NGTCP2_PV_FLAG_FALLBACK_PRESENT) &&
+      !(pv->flags & NGTCP2_PV_FLAG_DONT_RETIRE_FALLBACK) &&
       pv->fallback_dcid.cid.datalen &&
       pv->fallback_dcid.seq != conn->dcid.current.seq &&
       pv->fallback_dcid.seq != pv->dcid.seq) {
@@ -6251,9 +6255,9 @@ static int conn_recv_path_response(ngtcp2_conn *conn, const ngtcp2_pkt_hd *hd,
   }

   if (pv->flags & NGTCP2_PV_FLAG_FALLBACK_PRESENT) {
-    /* Unset the flag bit so that conn_stop_pv does not retire
+    /* Set the flag bit so that conn_stop_pv does not retire
        DCID. */
-    pv->flags &= (uint8_t)~NGTCP2_PV_FLAG_FALLBACK_PRESENT;
+    pv->flags |= NGTCP2_PV_FLAG_DONT_RETIRE_FALLBACK;
   }

   rv = conn_stop_pv(conn, ts);
@@ -8983,9 +8987,9 @@ static int conn_recv_non_probing_pkt_on_new_path(ngtcp2_conn *conn,
     if (conn->pv && (conn->pv->flags & NGTCP2_PV_FLAG_FALLBACK_PRESENT)) {
       ngtcp2_pv_set_fallback(pv, &conn->pv->fallback_dcid,
                              conn->pv->fallback_pto);
-      /* Unset the flag bit so that conn_stop_pv does not retire
+      /* Set the flag bit so that conn_stop_pv does not retire
          DCID. */
-      conn->pv->flags &= (uint8_t)~NGTCP2_PV_FLAG_FALLBACK_PRESENT;
+      conn->pv->flags |= NGTCP2_PV_FLAG_DONT_RETIRE_FALLBACK;
     } else if (!pref_addr_migration) {
       ngtcp2_pv_set_fallback(pv, &conn->dcid.current, pto);
     }
diff --git a/lib/ngtcp2_pv.h b/lib/ngtcp2_pv.h
index 28bdf722..29c75205 100644
--- a/lib/ngtcp2_pv.h
+++ b/lib/ngtcp2_pv.h
@@ -77,11 +77,15 @@ void ngtcp2_pv_entry_init(ngtcp2_pv_entry *pvent,
    path validation fails, then fallback to them.  If path validation
    succeeds, the fallback Destination Connection ID is retired if it
    is not zero length, and does not equal to the current Destination
-   Connection ID. */
+   Connection ID, and NGTCP2_PV_FLAG_DONT_RETIRE_FALLBACK is not
+   set. */
 #define NGTCP2_PV_FLAG_FALLBACK_PRESENT 0x04U
 /* NGTCP2_PV_FLAG_PREFERRED_ADDR indicates that client is migrating to
    server's preferred address.  This flag is only used by client. */
 #define NGTCP2_PV_FLAG_PREFERRED_ADDR 0x10U
+/* NGTCP2_PV_FLAG_DONT_RETIRE_FALLBACK prevents the fallback
+   Destination Connection ID from being retired. */
+#define NGTCP2_PV_FLAG_DONT_RETIRE_FALLBACK 0x20U

 typedef struct ngtcp2_pv ngtcp2_pv;

diff --git a/tests/ngtcp2_conn_test.c b/tests/ngtcp2_conn_test.c
index 74ece3b5..b268c371 100644
--- a/tests/ngtcp2_conn_test.c
+++ b/tests/ngtcp2_conn_test.c
@@ -11054,6 +11054,165 @@ void test_ngtcp2_conn_server_path_validation(void) {
   assert_true(conn->dcid.current.flags & NGTCP2_DCID_FLAG_PATH_VALIDATED);

   ngtcp2_conn_del(conn);
+
+  /* Client migrates back to the original path while server is doing
+     path validation to the new path. */
+  setup_default_server(&conn);
+  ngtcp2_tpe_init_conn(&tpe, conn);
+
+  /* This will send NEW_CONNECTION_ID frames */
+  spktlen = ngtcp2_conn_write_pkt(conn, NULL, NULL, buf, sizeof(buf), ++t);
+
+  assert_ptrdiff(0, <, spktlen);
+  assert_size(1, <, ngtcp2_ksl_len(&conn->scid.set));
+
+  /* A remote endpoint did not change DCID */
+  frs[0].ping.type = NGTCP2_FRAME_PING;
+  frs[1].padding = (ngtcp2_padding){
+    .type = NGTCP2_FRAME_PADDING,
+    .len = 1200,
+  };
+
+  pktlen = ngtcp2_tpe_write_1rtt(&tpe, buf, sizeof(buf), frs, 2);
+
+  rv = ngtcp2_conn_read_pkt(conn, &new_path1.path, NULL, buf, pktlen, ++t);
+
+  assert_int(0, ==, rv);
+  assert_not_null(conn->pv);
+
+  ngtcp2_path_storage_zero(&dst_path);
+  spktlen =
+    ngtcp2_conn_write_pkt(conn, &dst_path.path, NULL, buf, sizeof(buf), ++t);
+
+  assert_ptrdiff(NGTCP2_MAX_UDP_PAYLOAD_SIZE, <=, spktlen);
+  assert_size(0, <, ngtcp2_ringbuf_len(&conn->pv->ents.rb));
+  assert_true(ngtcp2_path_eq(&new_path1.path, &dst_path.path));
+
+  /* Client migrates back to the original path. */
+  frs[0].ping.type = NGTCP2_FRAME_PING;
+  frs[1].padding = (ngtcp2_padding){
+    .type = NGTCP2_FRAME_PADDING,
+    .len = 1200,
+  };
+
+  tpe.dcid = conn->oscid;
+
+  pktlen = ngtcp2_tpe_write_1rtt(&tpe, buf, sizeof(buf), frs, 2);
+
+  rv = ngtcp2_conn_read_pkt(conn, &null_path.path, NULL, buf, pktlen, ++t);
+
+  assert_int(0, ==, rv);
+  assert_null(conn->pv);
+  assert_true(ngtcp2_path_eq(&null_path.path, &conn->dcid.current.ps.path));
+  assert_true(ngtcp2_cid_eq(&(ngtcp2_cid)make_dcid(), &conn->dcid.current.cid));
+  assert_true(conn->dcid.current.flags & NGTCP2_DCID_FLAG_PATH_VALIDATED);
+  assert_size(0, ==, ngtcp2_dcidtr_retired_len(&conn->dcid.dtr));
+
+  ngtcp2_conn_del(conn);
+
+  /* DCID is not retired if it is used as fallback in new path
+     validation */
+  setup_default_server(&conn);
+  ngtcp2_tpe_init_conn(&tpe, conn);
+
+  /* This will send NEW_CONNECTION_ID frames */
+  spktlen = ngtcp2_conn_write_pkt(conn, NULL, NULL, buf, sizeof(buf), ++t);
+
+  assert_ptrdiff(0, <, spktlen);
+  assert_size(1, <, ngtcp2_ksl_len(&conn->scid.set));
+
+  frs[0].new_connection_id = (ngtcp2_new_connection_id){
+    .type = NGTCP2_FRAME_NEW_CONNECTION_ID,
+    .seq = 1,
+    .cid = cid,
+    .token = token,
+  };
+  frs[1].new_connection_id = (ngtcp2_new_connection_id){
+    .type = NGTCP2_FRAME_NEW_CONNECTION_ID,
+    .seq = 2,
+    .cid =
+      {
+        .datalen = 4,
+        .data = {0x1F, 0x00, 0x00, 0x00},
+      },
+    .token = token,
+  };
+
+  pktlen = ngtcp2_tpe_write_1rtt(&tpe, buf, sizeof(buf), frs, 2);
+
+  rv = ngtcp2_conn_read_pkt(conn, &null_path.path, NULL, buf, pktlen, ++t);
+
+  assert_int(0, ==, rv);
+
+  /* A remote endpoint did not change DCID */
+  frs[0].ping.type = NGTCP2_FRAME_PING;
+  frs[1].padding = (ngtcp2_padding){
+    .type = NGTCP2_FRAME_PADDING,
+    .len = 1200,
+  };
+
+  pktlen = ngtcp2_tpe_write_1rtt(&tpe, buf, sizeof(buf), frs, 2);
+
+  rv = ngtcp2_conn_read_pkt(conn, &new_path1.path, NULL, buf, pktlen, ++t);
+
+  assert_int(0, ==, rv);
+  assert_not_null(conn->pv);
+
+  ngtcp2_path_storage_zero(&dst_path);
+  spktlen =
+    ngtcp2_conn_write_pkt(conn, &dst_path.path, NULL, buf, sizeof(buf), ++t);
+
+  assert_ptrdiff(NGTCP2_MAX_UDP_PAYLOAD_SIZE, <=, spktlen);
+  assert_size(0, <, ngtcp2_ringbuf_len(&conn->pv->ents.rb));
+  assert_true(ngtcp2_path_eq(&new_path1.path, &dst_path.path));
+
+  /* Client migrates to new path with new DCID */
+  frs[0].ping.type = NGTCP2_FRAME_PING;
+  frs[1].padding = (ngtcp2_padding){
+    .type = NGTCP2_FRAME_PADDING,
+    .len = 1200,
+  };
+
+  it = ngtcp2_ksl_begin(&conn->scid.set);
+
+  assert(!ngtcp2_ksl_it_end(&it));
+
+  new_cid = &(((ngtcp2_scid *)ngtcp2_ksl_it_get(&it))->cid);
+  tpe.dcid = *new_cid;
+
+  pktlen = ngtcp2_tpe_write_1rtt(&tpe, buf, sizeof(buf), frs, 2);
+
+  rv = ngtcp2_conn_read_pkt(conn, &new_path2.path, NULL, buf, pktlen, ++t);
+
+  assert_int(0, ==, rv);
+  assert_not_null(conn->pv);
+  assert_size(0, ==, ngtcp2_dcidtr_retired_len(&conn->dcid.dtr));
+
+  ngtcp2_path_storage_zero(&dst_path);
+  spktlen =
+    ngtcp2_conn_write_pkt(conn, &dst_path.path, NULL, buf, sizeof(buf), ++t);
+
+  assert_ptrdiff(NGTCP2_MAX_UDP_PAYLOAD_SIZE, <=, spktlen);
+  assert_size(0, <, ngtcp2_ringbuf_len(&conn->pv->ents.rb));
+  assert_true(ngtcp2_path_eq(&new_path2.path, &dst_path.path));
+
+  fr.path_response = (ngtcp2_path_response){
+    .type = NGTCP2_FRAME_PATH_RESPONSE,
+  };
+
+  pktlen = ngtcp2_tpe_write_1rtt(&tpe, buf, sizeof(buf), &fr, 1);
+  rv = ngtcp2_conn_read_pkt(conn, &new_path2.path, NULL, buf, pktlen, ++t);
+
+  assert_int(0, ==, rv);
+  assert_true(ngtcp2_path_eq(&new_path2.path, &conn->dcid.current.ps.path));
+  assert_size(0, ==, ngtcp2_dcidtr_retired_len(&conn->dcid.dtr));
+  assert_not_null(conn->pv);
+  assert_true(ngtcp2_path_eq(&null_path.path, &conn->pv->dcid.ps.path));
+  assert_false(conn->pv->flags & NGTCP2_PV_FLAG_FALLBACK_PRESENT);
+  assert_false(conn->pv->flags & NGTCP2_PV_FLAG_DONT_RETIRE_FALLBACK);
+  assert_true(conn->pv->flags & NGTCP2_PV_FLAG_DONT_CARE);
+
+  ngtcp2_conn_del(conn);
 }

 void test_ngtcp2_conn_client_connection_migration(void) {
